# CSE185 The Super Naïve Low Dimensional Embedding Clustering Method 
This is group 16 final project for CSE185. It follows the algorithm description in  the [original t-SNE paper](https://lvdmaaten.github.io/publications/papers/JMLR_2008.pdf) to implement t-SNE tool. The main application of TeeSnee is to cluster gene expression data of cells to reveal cell types. 

[![Python Version](https://img.shields.io/badge/python-3.6%20|%203.7%20|%203.8%20|%203.9%20|%203.10-blue.svg)](https://www.python.org/downloads/)
[![PyPI Version](https://img.shields.io/badge/pip-v23.1.2-green.svg)](https://pip.pypa.io/en/stable/installation/)
[![NumPy](https://img.shields.io/badge/NumPy-1.24-yellow.svg)](https://numpy.org/doc/stable/index.html)
[![pandas](https://img.shields.io/badge/pandas-2.0.0-orange.svg)](https://pandas.pydata.org/)
[![Matplotlib](https://img.shields.io/badge/Matplotlib-3.7.0-purple.svg)](https://matplotlib.org/)
[![scikit-learn](https://img.shields.io/badge/scikit--learn-1.2.2-red.svg)](https://scikit-learn.org/stable/)
[![Anndata](https://img.shields.io/badge/anndata-v0.9.1-brightgreen)](https://github.com/theislab/anndata)
[![Scanpy](https://img.shields.io/badge/scanpy-v2.5.0-yellow)](https://scapy.readthedocs.io/en/latest/introduction.html)



# Table of contents <a name="toc"></a >
- [Sample Data Download](#data)
- [Installation Instructions](#install)
- [Basic Usage](#usage)
- [Complete Usage](#instruction)
- [Benchmarking Information](#benchmarking)
- [Contributors](#credit)

# Sample Data Download <a name="data"></a>
[BACK TO TABLE OF CONTENTS](#toc)

Sample data (count matrix) taken from: 
https://www.10xgenomics.com/resources/datasets/human-brain-cancer-11-mm-capture-area-ffpe-2-standard  

Check file `data_processing.py` to copy the commands needed to process the data to get gene by cell matrix.

# Installation Instruction <a name="install"></a>
[BACK TO TABLE OF CONTENTS](#toc)

* Install `teesnee` program with the following command:
```
git clone https://github.com/m1ma0314/CSE185Group16_TeeSnee.git
```
```
cd CSE185Group16_TeeSnee
```
* Python Installation
[Click Me to Get Python](https://www.python.org/downloads/)

* Pip Installation 
```
python -m ensurepip --upgrade
python -m pip install --upgrade pip
```

* Installation requires the `numpy`, `pandas`, `matplotlib`, `scikit-learn` libraries to be installed. You can install these with `pip`:

```
pip install -r requirements.txt
```
* Change permissions of teesnee.py:

```
chmod 777 teesnee.py
```

# Basic Usage <a name="usage"></a>
[BACK TO TABLE OF CONTENTS](#toc)

The basic usage of `teesnee` is:
```
python teesnee.py [-p targer_perlexity] [-z ifzipped] [-o output] filename
```

To run `teesnee` on a small test example (using files in this repo):
```
python teesnee.py -p 100 -o ./ minimal_dataset.csv
```
# Complete usage instructions <a name="instruction"></a>
[BACK TO TABLE OF CONTENTS](#toc)

The only required input to `my_tsne` is a cell x gene matrix data file. Users may additionally specify the options below:
* `-p PERPLEXITY`, `--target_perplexity PERPLEXITY`: specify target perplexity. If specified, the `tsne` function will calculate similarities matrix based on specified perplexity value and generate t-SNE plot. Higher perplexity value is associated with tighter clusters in the final output plot. Otherwise, the `tsne` function use `perplexity=100` by default.
* `-z ifzipped`, `--zipped`: unzip dataset file if this argument is specified. By default, the datafile is viewed as unzipped and will be converted into matrix for further processing.
* `-o FILE`, `--output FILE`: Write output to file. By default, output is written to `tsneplot.png`

## File Format
The output t-SNE plot is in `png` format.

![tsne_plot](https://github.com/m1ma0314/CSE185Group16_tSNE/assets/97704603/d9983cb9-58d6-416f-905b-bc62e580fb7c)

# Benchmarking Information <a name='benchmarking'></a>
[BACK TO TABLE OF CONTENTS](#toc)

The benchmarking analysis is recorded in the folder `benchmarking`. 
Here is the time complexity plot comparison between teeSnee time complexity and scanpy’s t-SNE.

<img width="616" alt="Screen Shot 2023-06-09 at 10 28 31 PM" src="https://github.com/m1ma0314/CSE185Group16_TeeSnee/assets/97704603/117aa3de-2cb3-4860-bd39-a77a9dec65dd">

p.s. teeSnee still needs optimization for runtime TAT

# Contributors <a name='credit'></a>
[BACK TO TABLE OF CONTENTS](#toc)

This repository was generated by Annabelle Coles and Mijia Ma, under the guidance of the [original t-SNE paper](https://lvdmaaten.github.io/publications/papers/JMLR_2008.pdf) and with inspiration from the article [t-SNE from scratch](https://towardsdatascience.com/t-sne-from-scratch-ft-numpy-172ee2a61df7). 

Special thanks to CSE185 Professor Dr. Melissa Gymrek, TA Ryan Eveloff, TA Luisa Amaral, TA Himanshu.

Please submit a pull request with any corrections or suggestions. Your suggestions matter a lot to us <3




